### PR TITLE
Core/PacketIO: Banksystem

### PR DESCRIPTION
**Pull request by Luzifix**
Thx @gigi1237 for search Opcodes
Thx @Nayd to help with InvUpdate & InvItem Structure

### CODE REVIEW

File: src/server/game/Handlers/BankHandler.cpp [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    TC_LOG_DEBUG("network", "WORLD: CMSG_AUTOSTORE_BANK_ITEM");
+    TC_LOG_DEBUG("network", "STORAGE: receive bag = %u, slot = %u", packet.Bag, packet.Slot);
+
+    if (!CanUseBank())
+    {
+        TC_LOG_ERROR("network", "WORLD: HandleAutoStoreBankItemOpcode - Unit (%s) not found or you can't interact with him.", m_currentBankerGUID.ToString().c_str());
+        return;
+    }
+
+    Item* pItem = _player->GetItemByPos(packet.Bag, packet.Slot);
```
Comment (DDuarte) [inline src/server/game/Handlers/BankHandler.cpp]: rename to `item`

File: src/server/game/Handlers/BankHandler.cpp [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    TC_LOG_DEBUG("network", "WORLD: CMSG_AUTOBANK_ITEM");
+    TC_LOG_DEBUG("network", "STORAGE: receive bag = %u, slot = %u", packet.Bag, packet.Slot);
+
+    if (!CanUseBank())
+    {
+        TC_LOG_ERROR("network", "WORLD: HandleAutoBankItemOpcode - Unit (%s) not found or you can't interact with him.", m_currentBankerGUID.ToString().c_str());
+        return;
+    }
+
+    Item* pItem = _player->GetItemByPos(packet.Bag, packet.Slot);
```
Comment (DDuarte) [inline src/server/game/Handlers/BankHandler.cpp]: rename to `item`

File: src/server/game/Handlers/BankHandler.cpp [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+#include "NPCPackets.h"
+#include "ObjectMgr.h"
+#include "Opcodes.h"
+#include "Player.h"
+#include "WorldPacket.h"
+#include "WorldSession.h"
+
+void WorldSession::HandleAutoBankItemOpcode(WorldPackets::Bank::AutoBankItem& packet)
+{
+    TC_LOG_DEBUG("network", "WORLD: CMSG_AUTOBANK_ITEM");
```
Comment (DDuarte) [inline src/server/game/Handlers/BankHandler.cpp]: remove this line

File: src/server/game/Handlers/BankHandler.cpp [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    _player->RemoveItem(packet.Bag, packet.Slot, true);
+    _player->ItemRemovedQuestCheck(pItem->GetEntry(), pItem->GetCount());
+    _player->BankItem(dest, pItem, true);
+}
+
+void WorldSession::HandleBankerActivateOpcode(WorldPackets::NPC::Hello& packet)
+{
+    TC_LOG_DEBUG("network", "WORLD: Received CMSG_BANKER_ACTIVATE");
```
Comment (DDuarte) [inline src/server/game/Handlers/BankHandler.cpp]: remove this line

File: src/server/game/Handlers/BankHandler.cpp [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        GetPlayer()->RemoveAurasByType(SPELL_AURA_FEIGN_DEATH);
+
+    // set currentBankerGUID for other bank action
+
+    SendShowBank(packet.Unit);
+}
+
+void WorldSession::HandleAutoStoreBankItemOpcode(WorldPackets::Bank::AutoStoreBankItem& packet)
+{
+    TC_LOG_DEBUG("network", "WORLD: CMSG_AUTOSTORE_BANK_ITEM");
```
Comment (DDuarte) [inline src/server/game/Handlers/BankHandler.cpp]: remove this line

File: src/server/game/Handlers/BankHandler.cpp [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+        }
+
+        _player->RemoveItem(packet.Bag, packet.Slot, true);
+        _player->BankItem(dest, pItem, true);
+    }
+}
+
+void WorldSession::HandleBuyBankSlotOpcode(WorldPackets::Bank::BuyBankSlot& packet)
+{
+    TC_LOG_DEBUG("network", "WORLD: CMSG_BUY_BANK_SLOT");
```
Comment (DDuarte) [inline src/server/game/Handlers/BankHandler.cpp]: remove this line

File: src/server/game/Handlers/BankHandler.cpp [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+        _player->RemoveItem(packet.Bag, packet.Slot, true);
+        _player->BankItem(dest, pItem, true);
+    }
+}
+
+void WorldSession::HandleBuyBankSlotOpcode(WorldPackets::Bank::BuyBankSlot& packet)
+{
+    TC_LOG_DEBUG("network", "WORLD: CMSG_BUY_BANK_SLOT");  
+
+    WorldPacket data(SMSG_BUY_BANK_SLOT_RESULT, 4);
```
Comment (DDuarte) [inline src/server/game/Handlers/BankHandler.cpp]: convert to the new packet structure, it's an easy one

File: src/server/game/Handlers/BankHandler.cpp [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+    _player->UpdateAchievementCriteria(ACHIEVEMENT_CRITERIA_TYPE_BUY_BANK_SLOT);
+}
+
+void WorldSession::SendShowBank(ObjectGuid guid)
+{
+    m_currentBankerGUID = guid;
+    WorldPackets::NPC::ShowBank packet;
+    packet.Guid = guid;
+    SendPacket(packet.Write());
+}
```
Comment (DDuarte) [inline src/server/game/Handlers/BankHandler.cpp]: add an empty line at the end of the file

File: src/server/game/Server/Packets/BankPackets.cpp [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+void WorldPackets::Bank::AutoStoreBankItem::Read()
+{
+    _worldPacket >> Inv >> Bag >> Slot;
+}
+
+void WorldPackets::Bank::BuyBankSlot::Read()
+{
+    _worldPacket >> Guid;
+}
```
Comment (DDuarte) [inline src/server/game/Server/Packets/BankPackets.cpp]: add an empty line at the end of the file

File: src/server/game/Server/Packets/BankPackets.h [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+#include "Packet.h"
+#include "ObjectGuid.h"
+#include "WorldSession.h"
+
+namespace WorldPackets
+{
+    namespace Bank
+    {
+        struct InvUpdate
```
Comment (joschiwald) [inline src/server/game/Server/Packets/BankPackets.h]: don't duplicate this struct, use it from WorldPacket::Item namespace

Comment (Luzifix) [inline src/server/game/Server/Packets/BankPackets.h]: If i change it i become this error https://nufla.de/kYLch.png =/ you have a idea how i can slove this error ?

Comment (joschiwald) [inline src/server/game/Server/Packets/BankPackets.h]: remove the new bytebuffer overloads, already implemented in itempackets

Comment (Luzifix) [inline src/server/game/Server/Packets/BankPackets.h]: thx, it works now

File: src/server/game/Server/Packets/BankPackets.cpp (hunk 9 above)
Comment (Luzifix) [inline src/server/game/Server/Packets/BankPackets.cpp]: added

**Comment by Luzifix** [comment]
anyone other test it ?

**Pull request merged**